### PR TITLE
PXB-3079 : Prepare skips rollback on encrypted tables and completes s…

### PR DESCRIPTION
**Pull request by satya-bodapati**
…uccessfully if keyring plugin is not loaded

https://jira.percona.com/browse/PXB-3079

Problem:
--------
If the backup contains an encrypted table and xtrabackup during prepare is not configured correctly, i.e. either keyring_file plugin is not passed to xtrabackup, or the keyring_file_data is not pointing to a valid keyring file, xtrabackup doesn't abort.

If there is a rollback on such encrypted tables, it skips rollback. The table is considered as dropped because it was not loaded to the cache. Encryption errors are ignored when the tablespace is loaded into the cache.

Fix:
----
If xtrabackup cannot load an encrypted tablespace during prepare, abort with an error.

### CODE REVIEW

**Comment by satya-bodapati** [comment]
https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/295/

**Comment by altmannmarcelo** [comment]
@satya-bodapati, what about single table backups? 
We might have a lot of missing tables. 
This is meant for fixing an issue with encryption but the fix is on general code path.
Will this patch break single table backup functionality?

**Comment by satya-bodapati** [comment]
> @satya-bodapati, what about single table backups? We might have a lot of missing tables. Will this patch break single table backup functionality?

No marce, for the IBDs that are missing, we will not see them. so no issues.

**Comment by altmannmarcelo** [comment]
@satya-bodapati Do you have an example of an xtrabackup output exploring the new code path? I would like to have a look on the output.

**Comment by satya-bodapati** [comment]
Although fil_open_for_xtrabackup() is used by backup and prepare, only the prepare caller aborts.

**Comment by satya-bodapati** [comment]
> @satya-bodapati Do you have an example of an xtrabackup output exploring the new code path? I would like to have a look on the output.

```
023-08-03T18:05:03.074433+01:00 0 [ERROR] [MY-012226] [InnoDB] Encryption information in datafile: ./test/t1.ibd can't be decrypted, please confirm that keyring is loaded.
2023-08-03T18:05:03.074467+01:00 0 [Note] [MY-011825] [Xtrabackup] Failed to decrypt table ./test/t1.ibd with space id 8. Will check if encrytion key has been parsed at the end of backup.
2023-08-03T18:05:03.074500+01:00 0 [ERROR] [MY-011825] [Xtrabackup] Unable to open tablespace 'test/t1.ibd'. InnoDB DB_ error code is 113

```

**Comment by altmannmarcelo** [comment]
@satya-bodapati As per slack conversation, we should try to only abort the backup if undo entries belong to missing encryption tables.

**Comment by altmannmarcelo** [comment]
@satya-bodapati, please also change the target branch to be the release branch if you want this on 8.0.34

**Comment by altmannmarcelo** [comment]
@satya-bodapati FYI: clang format is broken on your commit

**Comment by satya-bodapati** [comment]
Yes @altmannmarcelo , I fixed it locally. Will push it along with other changes

**Comment by satya-bodapati** [comment]
@altmannmarcelo updated the patch and added new testcase

File: share/messages_to_error_log.txt [hunk 1]
```diff
@@ -12267,6 +12267,9 @@ ER_KEYRING_COMPONENT_CONFIG_PARSE_FAILED
 ER_KEYRING_COMPONENT_KMS_ERROR
   eng "KMS communication error: %s"
 
+ER_KEYRING_NOT_LOADED
```
Comment (altmannmarcelo) [inline share/messages_to_error_log.txt]: Can we convert this to an inline comment? We use the same message multiple times but usually avoid having PXB custom messages on shared files.

File: storage/innobase/fil/fil0fil.cc [hunk 2]
```diff
@@ -2257,13 +2279,20 @@ void Tablespace_files::open_ibds_slice(const Const_iter &start,
...
   for (auto it = start; it != end; ++it) {
     for (auto name : it->second) {
-      dberr_t err = fil_open_for_xtrabackup(m_dir.path() + name,
-                                            name.substr(0, name.length() - 4));
-
-      if (err != DB_SUCCESS && err != DB_TABLESPACE_EXISTS &&
-          err != DB_TABLESPACE_NOT_FOUND && err != DB_TABLESPACE_DELETED) {
-        xb::info() << "Unable to open tablespace " << SQUOTE(name)
-                   << ". Possibly dropped. InnoDB DB_ error code is " << err;
+      // result <dberr_t, space_id_t>
```
Comment (altmannmarcelo) [inline storage/innobase/fil/fil0fil.cc]: cosmetic, use `/* */`

File: storage/innobase/fil/fil0fil.cc [hunk 3]
```diff
@@ -2257,13 +2279,20 @@ void Tablespace_files::open_ibds_slice(const Const_iter &start,
...
+      auto [err, space_id] = fil_open_for_xtrabackup(
+          m_dir.path() + name, name.substr(0, name.length() - 4));
+
+      // DB_TABLESPACE_EXISTS is allowed because it means a tablespace is
+      // already opened by redo
```
Comment (altmannmarcelo) [inline storage/innobase/fil/fil0fil.cc]: same here and inside the if, use multi-line comment

File: storage/innobase/xtrabackup/src/xb_dict.cc [hunk 4]
```diff
@@ -228,7 +228,26 @@ b - std::vector<dict_table_t*>, empty on errors */
...
+    if (err != DB_ERROR_UNSET) {
+      std::string tablespace_name;
+      bool found = fil_system_get_file_by_space_id(space_id, tablespace_name);
+      xb::error() << "Tablespace " << (found ? tablespace_name : "")
+                  << " with space_id " << space_id
+                  << " is required for transaction rollback but it was not"
+                  << " loaded because of DB_ error " << err;
+
+      if (err == DB_INVALID_ENCRYPTION_META) {
+        xb::error(ER_KEYRING_NOT_LOADED);
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xb_dict.cc]: I would keep the error only here, and not duplicate on the other places.

File: storage/innobase/fil/fil0fil.cc [hunk 5]
```diff
@@ -2257,13 +2279,20 @@ void Tablespace_files::open_ibds_slice(const Const_iter &start,
...
+      if (err != DB_SUCCESS && err != DB_TABLESPACE_EXISTS) {
+        if (space_id != SPACE_UNKNOWN) {
```
Comment (altmannmarcelo) [inline storage/innobase/fil/fil0fil.cc]: Any reason for the extra if here ? it could be done above

File: storage/innobase/fsp/fsp0file.cc [hunk 6]
```diff
@@ -691,11 +697,14 @@ dberr_t Datafile::validate_first_page(space_id_t space_id, lsn_t *flush_lsn,
         ut::free(m_encryption_iv);
         m_encryption_key = nullptr;
         m_encryption_iv = nullptr;
-        xb::info() << "Failed to decrypt table " << m_filepath << " with space"
-                   << " id " << m_space_id
-                   << ". Will check if encrytion key has been"
-                   << " parsed at the end of backup.";
-        invalid_encrypted_tablespace_ids.push_back(m_space_id);
+        if (srv_backup_mode) {
```
Comment (altmannmarcelo) [inline storage/innobase/fsp/fsp0file.cc:700]: Nice, this also fixes the bogus message during prepare.

File: storage/innobase/xtrabackup/src/xb_dict.cc [hunk 7]
```diff
@@ -228,7 +228,26 @@ b - std::vector<dict_table_t*>, empty on errors */
...
+    dberr_t err = fil_xb_get_tablespace_error(space_id);
+    /* Check if tablespace errored out when it was loaded. If yes, we abort
+    the prepare. If no, we consider the tablespace as dropped */
+    if (err != DB_ERROR_UNSET) {
+      std::string tablespace_name;
+      bool found = fil_system_get_file_by_space_id(space_id, tablespace_name);
+      xb::error() << "Tablespace " << (found ? tablespace_name : "")
+                  << " with space_id " << space_id
+                  << " is required for transaction rollback but it was not"
+                  << " loaded because of DB_ error " << err;
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xb_dict.cc:242]: Do we print DB_ERROR or INT here ?

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
Thanks for re-working it to only abort if we have rollback on the said tablespace.
Few minor comments. Since we basically re-worked a good part of the logic, can we have another Jenkins run?

File: storage/innobase/xtrabackup/src/xb_dict.cc (hunk 7 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/xb_dict.cc:242]: unfortunately it is a number. We dont have a string mapping from the c style enum.

File: storage/innobase/xtrabackup/src/xb_dict.cc (hunk 4 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/xb_dict.cc]: I felt that other places related to encryption errors deserve better messages. The user has no clue when he sees "invalid redo" , "invalid encryption headers," etc.

These can be caused by lack of keyring and keyring configurations.

File: storage/innobase/fsp/fsp0file.cc (hunk 6 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0file.cc:700]: there is other secret intention for doing this change.

Tablespace loading is parallel (open_ibds_slice). And each thread can insert into this invalid_encrypted_tabelsapce_ids parallel and there is no mutex protection for this structure. Avoiding a possible race condition

File: storage/innobase/fil/fil0fil.cc (hunk 3 above)
Comment (satya-bodapati) [reply storage/innobase/fil/fil0fil.cc]: Didn't know that multi line comments with // are not recommended. May be just for consistency reasons. But it is convenient when writing them ;)

File: storage/innobase/fsp/fsp0file.cc (hunk 6 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0file.cc:700]: Just checked code, looks like this is a problem for backup too? :thinking:

Comment (altmannmarcelo) [reply storage/innobase/fsp/fsp0file.cc:700]: IS the initial scan done in parallel on --backup too now ?

File: storage/innobase/fil/fil0fil.cc (hunk 3 above)
Comment (altmannmarcelo) [reply storage/innobase/fil/fil0fil.cc]: We normally use `//` for end of line comments, but multi-line read better with `/* */` - just cosmetic change

File: storage/innobase/fsp/fsp0file.cc (hunk 6 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0file.cc:700]: yes, here is the code flow
```
Tablespace_dirs::scan->Tablespace_dirs::open_ibd->fil_open_for_xtrabackup->file.validate_first_page->invalid_encrypted_tablespace_ids

Tablespace_dirs::scan
  par_for(PFS_NOT_INSTRUMENTED, ibd_files, n_threads, check, &m, &unique,
            &duplicates);
  }

```

Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0file.cc:700]: but we have to test and see if there are more than 1 threads or not

File: storage/innobase/fil/fil0fil.cc (hunk 5 above)
Comment (satya-bodapati) [reply storage/innobase/fil/fil0fil.cc]: can be combined. done

File: share/messages_to_error_log.txt (hunk 1 above)
Comment (satya-bodapati) [reply share/messages_to_error_log.txt]: Done

File: storage/innobase/fsp/fsp0file.cc (hunk 6 above)
Comment (altmannmarcelo) [reply storage/innobase/fsp/fsp0file.cc:700]: Can you open a separate ticket for this ?

**Review by altmannmarcelo — APPROVED** [review]
LGTM. Please get the last round of jenkins and squash the commits

**Comment by altmannmarcelo** [comment]
@satya-bodapati Don't forget to target this PR to RB.

**Comment by satya-bodapati** [comment]
> @satya-bodapati Don't forget to target this PR to RB.

Yeah, will rebase to release branch

**Comment by satya-bodapati** [comment]
> LGTM. Please get the last round of jenkins and squash the commits

Will squash and submit to Jenkins

File: storage/innobase/fsp/fsp0file.cc (hunk 6 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0file.cc:700]: Done marce: https://jira.percona.com/browse/PXB-3118

**Comment by satya-bodapati** [comment]
https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/298/

**Comment by satya-bodapati** [comment]
> https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/298/

https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-multijob/299/

**Pull request merged**